src/plugins/knex.js

```diff
@@ -2,22 +2,145 @@
+      const builder = original.apply(this, arguments)
+      builder[spanSymbol] = span
+      return builder
+    }
+  }
+}
+
+function createWrapRunner (wrapper, tracer, config) {
+  return function wrapRunner (original) {
+    return function runnerWithTrace () {
+      const runner = original.apply(this, arguments)
+      wrapper.wrap(runner, 'query', createWrapRunnerQuery(tracer, config))
+      return runner
+    }
+  }
+}
+
+function createWrapRunnerQuery (tracer, config) {
+  return function wrapQuery (original) {
+    return function queryWithTrace (q) {
+      const scope = tracer.scope()
+      const childOf = this.builder[spanSymbol]
+
+      const tags = {
+        'component': 'knex',
+        'db.statement': q.sql
+      }
+      if (q.timeout !== undefined) {
+        tags.timeout = q.timeout
+      }
+
+      let spanName = 'knex.client.runner'
+      if (q.method !== undefined) {
+        spanName = `knex.client.runner.${q.method}`
+      }
+      const span = tracer.startSpan(spanName, {
+        childOf,
+        tags
+      })
+      setDBTags(this, span)
+
+      return scope.activate(span, () => {
+        try {
+          return original.apply(this, arguments)
+        } catch (e) {
+          throw addError(span, e)
+        } finally {
+          span.finish()
+        }
+      })
     }
   }
 }
 
-module.exports = [
-  createPatch('lib/query/builder.js'),
-  createPatch('lib/raw.js'),
-  createPatch('lib/schema/builder.js')
-]
+function addError (span, error) {
+  span.setTag('error', 'true')
+  span.log({
+    'error.type': error.name,
+    'error.msg': error.message,
+    'error.stack': error.stack
+  })
+  return error
+}
+
+function setDBTags (obj, span) {
+  if (obj.client && obj.client.config) {
+    const config = obj.client.config
+    if (config.client) {
+      span.setTag('db.type', config.client)
+    }
+    if (config.connection && config.connection.user) {
+      span.setTag('db.user', config.connection.user)
+    }
+  }
+}
+
+function patchKnex (version, basePath) {
+  return [
+    {
+      name: 'knex',
+      versions: version,
+      file: `${basePath}/client.js`,
+      patch (Client, tracer, config) {
+        this.wrap(Client.prototype, 'queryBuilder', createWrapBuilder(tracer, config))
+        this.wrap(Client.prototype, 'schemaBuilder', createWrapBuilder(tracer, config))
+        this.wrap(Client.prototype, 'raw', createWrapBuilder(tracer, config))
+        this.wrap(Client.prototype, 'runner', createWrapRunner(this, tracer, config))
+      },
+      unpatch (Client) {
+        this.unwrap(Client.prototype, 'runner')
+        this.unwrap(Client.prototype, 'raw')
+        this.unwrap(Client.prototype, 'schemaBuilder')
+        this.unwrap(Client.prototype, 'queryBuilder')
+      }
+    },
+    // Not sure if patching the following `then` methods help in anyway now but I'm leaving them
+    // in as they did solve some problems in the past and they don't seem to have any adverse effects
```

Comment:
I believe the intention is to ensure scope propagation where bluebird and native promise instrumentation fail.  What about your changes make this unnecessary?

Reply:
My changes rely on the first call to a query builder to be able to extract context from the trace scope. Post that it passes the context on the builder itself. I removed these patches and wasn't didn't see any regressions but wasn't confident enough that they wouldn't be needed.

Reply:
Ah, since only query builders needs the initial context from async scope, I can only patch that and get rid of everything else.

Reply:
On second thought, I'm removing these as I'm unable to prove that they are actually needed. I couldn't break any tests or example apps without these patches. We can always add them back if we find a bug later.